streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/KTableRepartitionerProcessorSupplier.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.internals.foreignkeyjoin;
+
+import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.common.utils.Murmur3;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.streams.kstream.ValueMapper;
+import org.apache.kafka.streams.kstream.internals.Change;
+import org.apache.kafka.streams.processor.AbstractProcessor;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+
+import java.security.NoSuchAlgorithmException;
+import java.util.Random;
+
+public class KTableRepartitionerProcessorSupplier<K, KO, V> implements ProcessorSupplier<K, Change<V>> {
+
+    private final ValueMapper<V, KO> mapper;
+    private final Serializer<V> valueSerializer;
+
+    public KTableRepartitionerProcessorSupplier(final ValueMapper<V, KO> extractor,
+                                                final Serializer<V> valueSerializer) {
+        this.mapper = extractor;
+        this.valueSerializer = valueSerializer;
+    }
+
+    @Override
+    public Processor<K, Change<V>> get() {
+        return new UnbindChangeProcessor();
+    }
+
+    private class UnbindChangeProcessor extends AbstractProcessor<K, Change<V>> {
+        
+        @Override
+        public void init(final ProcessorContext context) {
+            super.init(context);
+        }
+
+        @Override
+        public void process(final K key, final Change<V> change) {
+            long[] nullHash = Murmur3.hash128(new byte[]{});
```

Comment:
The key can be anything. The issue is that Murmur3 cannot hash a null to a unique sentinel that wouldn't also be available by hashing another value. So any value we choose will not work.

That being said, I could do something like (you mentioned using a bit in another comment):
Hash is null: `{1-bit nullhash = true}{remaining-serialized-data}`.
Hash is notNull: `{1-bit nullhash = false}{2-bytes hash}{remaining-serialized-data}`.

The only thing I am rusty on is how the bit is handled in over-the-wire communication. Would it not simply end up being stored as a full byte anyways? In this case, we're adding a full byte to each message just to indicate if the hash is actually null, and this is in both directions. However, for correctness sake I believe we should do this as I don't think there is any safe sentinel we can use.

If we want to save space, we have two more options. 
1) Fold this in with the Version byte, and only give Version 7-bits.  {1-bit-nullhash}{7-bit-version}
2) Can also switch to Hash64 (50% chance of collision in ~4 Billion events) instead of Hash128 (50% chance of collision in ~1.844674407e19).

For the sake of moving this along, I will just include a single-bit of isNullHash and take it from the Version byte. This should satisfy our needs.

Reply:
I folded all this in, the 7-bit version and the 1-bit and the 1 bit "isNull". I'll resolve this one now and if it's no good for whatever reason then just open another.